Job did not finish on cluster with more than 2 members if Python code failed

In case when python code fails in pipeline then job did not finish on cluster with 2 and more members. It works on single node cluster. 

Job is restarted every ~2minutes and 13seconds and following Exception can be seen in server log:
```
ERROR [c.h.j.i.o.StartExecutionOperation] [hz.loving_sutherland.jet.cooperative.thread-2] - [10.212.134.151]:5702 [jet] [4.0-SNAPSHOT] Exception during callback
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.spi.impl.operationservice.impl.responses.ErrorResponse'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:157)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:133)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:124)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.send(OutboundResponseHandler.java:114)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.sendResponse(OutboundResponseHandler.java:94)
	at com.hazelcast.spi.impl.operationservice.Operation.sendResponse(Operation.java:470)
	at com.hazelcast.jet.impl.operation.AsyncOperation.doSendResponse(AsyncOperation.java:76)
	at com.hazelcast.jet.impl.operation.AsyncOperation.lambda$run$0(AsyncOperation.java:56)
	at com.hazelcast.jet.impl.util.ExceptionUtil.lambda$withTryCatch$0(ExceptionUtil.java:180)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:59)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:458)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:392)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:378)
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.jet.JetException'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:253)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:378)
	at com.hazelcast.spi.impl.operationservice.impl.responses.ErrorResponse.writeData(ErrorResponse.java:49)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:244)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.write(DataSerializableSerializer.java:51)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	... 19 more
Caused by: java.io.NotSerializableException: io.grpc.Status
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:441)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:441)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:441)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:115)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:108)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:251)
	... 25 more
``` 

How to reproduce:
Use following python script named `failing.py`:
```
def handle(input_list):
    assert 1 == 2
    return ['echo-%s' % i for i in input_list]
```
Use follwing pipeline:
```
        PythonServiceConfig cfg = new PythonServiceConfig()
                .setBaseDir(PATH_TO_BASE_DIR)
                .setHandlerModule("failing")
                .setHandlerFunction("handle");
        Pipeline p = Pipeline.create();
        p.readFrom(TestSources.items("1"))
                .apply(mapUsingPythonBatch(cfg)).setLocalParallelism(2)
                .writeTo(Sinks.logger());
        jet.newJob(p).join();
```

--- a/extensions/python/src/main/java/com/hazelcast/jet/python/PythonService.java
+++ b/extensions/python/src/main/java/com/hazelcast/jet/python/PythonService.java
@@ -132,7 +132,7 @@ final class PythonService {
         public void onError(Throwable e) {
             try {
                 if (e instanceof StatusException || e instanceof StatusRuntimeException) {
-                    // not seriazliable exceptions
+                    // not serializable exceptions
                     e = new JetException(ExceptionUtil.stackTraceToString(e));
                 }
 

--- a/extensions/python/src/main/java/com/hazelcast/jet/python/PythonService.java
+++ b/extensions/python/src/main/java/com/hazelcast/jet/python/PythonService.java
@@ -16,6 +16,7 @@
 package com.hazelcast.jet.python;
 
 import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.impl.util.ExceptionUtil;
 import com.hazelcast.jet.pipeline.ServiceFactory;
 import com.hazelcast.jet.python.grpc.InputMessage;
 import com.hazelcast.jet.python.grpc.InputMessage.Builder;
@@ -25,6 +26,8 @@ import com.hazelcast.jet.python.grpc.OutputMessage;
 import com.hazelcast.logging.ILogger;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import io.grpc.ManagedChannel;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
 import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
 import io.grpc.stub.StreamObserver;
 
@@ -128,6 +131,11 @@ final class PythonService {
         @Override
         public void onError(Throwable e) {
             try {
+                if (e instanceof StatusException || e instanceof StatusRuntimeException) {
+                    // not seriazliable exceptions
+                    e = new JetException(ExceptionUtil.stackTraceToString(e));
+                }
+
                 exceptionInOutputObserver = e;
                 for (CompletableFuture<List<String>> future; (future = futureQueue.poll()) != null;) {
                     future.completeExceptionally(e);

--- a/extensions/python/src/test/java/com/hazelcast/jet/python/PythonInitCleanupTest.java
+++ b/extensions/python/src/test/java/com/hazelcast/jet/python/PythonInitCleanupTest.java
@@ -32,7 +32,6 @@ import java.util.concurrent.CompletionException;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.BeforeClass;
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 
@@ -195,7 +194,6 @@ public class PythonInitCleanupTest extends SimpleTestInClusterSupport {
     }
 
     @Test
-    @Ignore("https://github.com/hazelcast/hazelcast-jet/issues/1995")
     public void cleanupExecutedIfPythonFailed() throws IOException {
         // Given
         installFileToBaseDir(FAILING_FUNCTION, "failing.py");

--- a/extensions/python/src/test/java/com/hazelcast/jet/python/PythonServiceTest.java
+++ b/extensions/python/src/test/java/com/hazelcast/jet/python/PythonServiceTest.java
@@ -49,7 +49,12 @@ public class PythonServiceTest extends SimpleTestInClusterSupport {
     private static final int ITEM_COUNT = 10_000;
     private static final String ECHO_HANDLER_FUNCTION =
             "def handle(input_list):\n" +
-            "    return ['echo-%s' % i for i in input_list]\n";
+ "    return ['echo-%s' % i for i in input_list]\n";
+
+    private static final String FAILING_FUNCTION
+            = "def handle(input_list):\n"
+            + "    assert 1 == 2\n"
+            + "    return ['echo-%s' % i for i in input_list]\n";
 
     private File baseDir;
 
@@ -261,6 +266,54 @@ public class PythonServiceTest extends SimpleTestInClusterSupport {
                 .setBaseDir(baseDir.toString())
                 .setHandlerModule("echo")
                 .setHandlerFunction("notExistsFunction");
+        Pipeline p = Pipeline.create();
+        StreamStage<String> stage = p.readFrom(TestSources.items("1")).addTimestamps(x -> 0, 0);
+
+        // When
+        stage.apply(mapUsingPython(cfg)).setLocalParallelism(2)
+                .writeTo(Sinks.logger());
+
+        // Then
+        try {
+            instance().newJob(p).join();
+            fail();
+        } catch (CompletionException ex) {
+            // expected
+        }
+    }
+
+    @Test
+    public void batchStage_mapUsingPythonException() throws IOException {
+        // Given
+        installFileToBaseDir(FAILING_FUNCTION, "failing.py");
+        PythonServiceConfig cfg = new PythonServiceConfig()
+                .setBaseDir(baseDir.toString())
+                .setHandlerModule("failing")
+                .setHandlerFunction("handle");
+        Pipeline p = Pipeline.create();
+        BatchStage<String> stage = p.readFrom(TestSources.items("1"));
+
+        // When
+        stage.apply(mapUsingPythonBatch(cfg)).setLocalParallelism(2)
+                .writeTo(Sinks.logger());
+
+        // Then
+        try {
+            instance().newJob(p).join();
+            fail();
+        } catch (CompletionException ex) {
+            // expected
+        }
+    }
+
+    @Test
+    public void streamStage_mapUsingPythonException() throws IOException {
+        // Given
+        installFileToBaseDir(FAILING_FUNCTION, "failing.py");
+        PythonServiceConfig cfg = new PythonServiceConfig()
+                .setBaseDir(baseDir.toString())
+                .setHandlerModule("failing")
+                .setHandlerFunction("handle");
         Pipeline p = Pipeline.create();
         StreamStage<String> stage = p.readFrom(TestSources.items("1")).addTimestamps(x -> 0, 0);
 

--- a/extensions/python/src/test/java/com/hazelcast/jet/python/PythonServiceTest.java
+++ b/extensions/python/src/test/java/com/hazelcast/jet/python/PythonServiceTest.java
@@ -49,12 +49,11 @@ public class PythonServiceTest extends SimpleTestInClusterSupport {
     private static final int ITEM_COUNT = 10_000;
     private static final String ECHO_HANDLER_FUNCTION =
             "def handle(input_list):\n" +
- "    return ['echo-%s' % i for i in input_list]\n";
+            "    return ['echo-%s' % i for i in input_list]\n";
 
     private static final String FAILING_FUNCTION
             = "def handle(input_list):\n"
-            + "    assert 1 == 2\n"
-            + "    return ['echo-%s' % i for i in input_list]\n";
+            + "    assert 1 == 2\n";
 
     private File baseDir;
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
@@ -25,12 +25,11 @@ import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.spi.impl.operationservice.ExceptionAction;
 import com.hazelcast.spi.impl.operationservice.Operation;
 
-import java.io.PrintWriter;
-import java.io.StringWriter;
 import java.util.concurrent.CompletableFuture;
 
 import static com.hazelcast.jet.impl.util.ExceptionUtil.isRestartableException;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.peel;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.stackTraceToString;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.withTryCatch;
 import static com.hazelcast.spi.impl.operationservice.ExceptionAction.THROW_EXCEPTION;
 
@@ -83,9 +82,7 @@ public abstract class AsyncOperation extends Operation implements IdentifiedData
                 Throwable ex = peel(e);
                 if (value instanceof Throwable && ex instanceof HazelcastSerializationException) {
                     // we got a non-serializable exception here
-                    StringWriter caughtStr = new StringWriter();
-                    ((Throwable) value).printStackTrace(new PrintWriter(caughtStr));
-                    sendResponse(new JetException(caughtStr.toString()));
+                    sendResponse(new JetException(stackTraceToString(ex)));
                 } else {
                     throw e;
                 }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
@@ -16,13 +16,17 @@
 
 package com.hazelcast.jet.impl.operation;
 
+import com.hazelcast.jet.JetException;
 import com.hazelcast.jet.impl.JetService;
 import com.hazelcast.jet.impl.JobCoordinationService;
 import com.hazelcast.jet.impl.execution.init.JetInitDataSerializerHook;
+import com.hazelcast.nio.serialization.HazelcastSerializationException;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.spi.impl.operationservice.ExceptionAction;
 import com.hazelcast.spi.impl.operationservice.Operation;
 
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.util.concurrent.CompletableFuture;
 
 import static com.hazelcast.jet.impl.util.ExceptionUtil.isRestartableException;
@@ -73,7 +77,19 @@ public abstract class AsyncOperation extends Operation implements IdentifiedData
             final JetService service = getService();
             service.getLiveOperationRegistry().deregister(this);
         } finally {
-            sendResponse(value);
+            try {
+                sendResponse(value);
+            } catch (Exception e) {
+                Throwable ex = peel(e);
+                if (value instanceof Throwable && ex instanceof HazelcastSerializationException) {
+                    // we got a non-serializable exception here
+                    StringWriter caughtStr = new StringWriter();
+                    ((Throwable) value).printStackTrace(new PrintWriter(caughtStr));
+                    sendResponse(new JetException(caughtStr.toString()));
+                } else {
+                    throw e;
+                }
+            }
         }
     }
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/AsyncOperation.java
@@ -81,7 +81,12 @@ public abstract class AsyncOperation extends Operation implements IdentifiedData
             } catch (Exception e) {
                 Throwable ex = peel(e);
                 if (value instanceof Throwable && ex instanceof HazelcastSerializationException) {
-                    // we got a non-serializable exception here
+                    // Sometimes exceptions are not serializable, for example on
+                    // https://github.com/hazelcast/hazelcast-jet/issues/1995.
+                    // When sending exception as a response and the serialization fails,
+                    // the response will not be sent and the operation will hang.
+                    // To prevent this from happening, replace the exception with
+                    // another exception that can be serialized.
                     sendResponse(new JetException(stackTraceToString(ex)));
                 } else {
                     throw e;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/ExceptionUtil.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/ExceptionUtil.java
@@ -40,6 +40,8 @@ import com.hazelcast.spi.exception.TargetNotMemberException;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Arrays;
 import java.util.List;
@@ -188,6 +190,13 @@ public final class ExceptionUtil {
     @Nonnull
     public static <T extends Throwable> RuntimeException sneakyThrow(@Nonnull Throwable t) throws T {
         throw (T) t;
+    }
+
+    @Nonnull
+    public static String stackTraceToString(Throwable t) {
+        StringWriter sw = new StringWriter();
+        t.printStackTrace(new PrintWriter(sw));
+        return sw.toString();
     }
 
     /**

